x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/DatafeedUpdate.java

```diff
@@ -169,14 +171,17 @@ public void writeTo(StreamOutput out) throws IOException {
             out.writeStringCollection(Collections.emptyList());
         }
         if (out.getVersion().before(Version.V_7_1_0)) {
-            out.writeOptionalNamedWriteable(lazyQueryParser.apply(query, id, new ArrayList<>()));
-            out.writeOptionalWriteable(lazyAggParser.apply(aggregations, id, new ArrayList<>()));
-        } else {
-            out.writeMap(query);
-            out.writeBoolean(aggregations != null);
-            if (aggregations != null) {
-                out.writeMap(aggregations);
+            out.writeBoolean(queryProvider != null);
```

Comment:
This is a breach of encapsulation for `writeOptionalNamedWriteable` - you're reproducing its internal logic here.  You'll get away with it in this case because its internal implementation is constrained by the need for BWC, but I think it would be better practice to keep using `writeOptionalNamedWriteable` as follows:

```
    out.writeOptionalNamedWriteable((queryProvider != null) ? queryProvider.getParsedQuery() : null);
```

And the same for `aggProvider` below if possible.